### DISCUSSION TITLE
USBIP Support for 7530

### DISCUSSION THREAD

**Discussion by MartB** [Freetz-NG/freetz-ng#884] [Allgemein]
Is it possible to use the 7530 as an usbip server for clients on linux?
At the moment its impossible to enable the usbip package because it was made for kernel 2.6 as it used to be out-of-tree.

What do i need to do to get usbip working on these devices?

**Comment by fda77** [comment]
Build the Kernelmodule and check if you could load it on the device

**Reply by MartB** [reply]
That would involve building the userspace linux-tools for usbip too and i dont think there is any facilty for building these tools included in the kernel source available right now?

**Reply by fda77** [reply]
Wasnt thew usbip for old devices?

**Comment by coldfix** [comment]
Hi, I'm also trying to setup usbip on my 7530 in case you're interested. I've [started a branch](https://github.com/Freetz-NG/freetz-ng/compare/master...coldfix:freetz-ng:usbip) on which I've managed to build the usbip kernel modules that are now integrated in the linux kernel mainline, as well as the usbip userspace tool. The latter required some nasty workarounds. I've already successfully deployed this on my box to connect a bluetooth dongle from my box's USB port to my PC.

This is still a proof-of-concept prototype with serious issues that is _not anywhere_ near ready for integration in freetz-ng (and might never be):

- fritzbox crashes on `usbip unbind -b <device>`
- also crashes when device is unplugged  (haven't done any investigation on either case so far)

Building the kernel modules that are distributed with the linux kernel seems to be unproblematic, the userspace tools are not:
- depends on  `libudev.so` which is not part of the kernel but is shipped with the fritzos image. I was able to use it with the following workarounds:
  - the `.mk` file links against the `.so` file in `build/original/filesystem/lib/`, but this works only if the freetz has already been built  (i.e. build once without usbip, then select usbip and run make again)
  - there is only a `libudev.so.1`, so can't you can't use `-ludev`. I therefore created a symlink named `libudev.so`. It may be possible to do specify `:libudev.so.1` instead of `udev` as dependency in `configure.ac` instead. **Edit: Confirmed, this works**
  - `libudev.h` is not shipped, so I've added a stripped-down version that includes only the symbols that are actually used by us. I have so far not checked whether these symbols are compatible with the ones defined by the shipped `libudev.so.1`
- I'm not sure what is the best way to create a package that builds a subfolder of the kernel source. I tried simply declaring the same source/hash/site as the kernel package does, but then I got some warnings and decided to simply copy over the `tools/usb/usbip` subfolder instead...
- the usbip userspace tools included in the linux 4.4.271 kernel source that is used for the 7530 does not build out of the box without patches due to multiple definition of `udev_context` symbol and misusage of `strncpy`
- necessity for these patches depends on the kernel version. They are not needed with newer kernels where this issue has been fixed. So, I guess, we would need patches per kernel version. I've seen that freetz-ng has this sort of patches, but this would need to live in the kernel build stuff, not with the usbip package, right?

I'd be thankful for any input :)

And less serious issues:
- I did not update the CGI webinterface (this should be relatively straight-forward I believe)
- I've added it next to the existing usbip package, but if its problems can be eradicated one can think whether it's better or not to replace it instead(?)
- (not sure why the package is not listed in the unstable section in the menuconfig?)
- so far building with `--with-tcp-wrappers=no`. It might be nice to enable if I `libwrap0` is available
- I think glib2 is not needed, but needs to be confirmed **Edit: `readelf -d` shows no dependency on glib2**
- side question: in the make rule e.g. `$(SUBMAKE) -C $(USBIP_LINUX_DIR)` I cannot use `-C $($(PKG)_LINUX_DIR)` as is done elsewhere, not sure why..

It might potentially be easier to fork the usbip userspace tools from the kernel and again maintain as a separate version..


Usage

```
## Commands for host (freetz):
    modprobe usbip-core
    modprobe usbip-host
    usbipd &
    usbip bind -b 1-1

## Commands for client
    sudo modprobe vhci_hcd
    sudo usbip attach -r HOSTIP -b BUSID
```

**Comment by fda77** [comment]
> fritzbox crashes on

maybe because of wrong libs? all freetz bins use /usr/lib/freetz/, nothing else! also freetz has own libc

> I'm not sure what is the best way to create a package that builds a subfolder of the kernel source.

i dont know, but maybe these package help you?

https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/wireguard-linux-compat/wireguard-linux-compat.mk
https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/wireguard/wireguard.mk
https://github.com/Freetz-NG/freetz-ng/blob/master/make/pkgs/fuse-kernel-module/fuse-kernel-module.mk

> but this would need to live in the kernel build stuff, not with the usbip package, right?

i think but down know...

https://github.com/Freetz-NG/freetz-ng/tree/master/make/kernel/patches/3.10.10x

could help

> I think glib2 is not needed, but needs to be confirmed

"ldd" or "readelf" show dependencies. 

> I cannot use `$($(PKG)_LINUX_DIR) `

i depends where / when evaluated! if unsure dont use `$(PKG)` :)


>  it's better or not to replace it instead(?)

yes, the old devices then no longer work. I've many packages with different versions, like `grep ABANDON  make/*/*/*.mk`


> not sure why the package is not listed 

in menuconfig type `/` and seatch for the symbol name, it shows dependencies even it is not visible

**Reply by fda77** [reply]
you added "usbip-core.ko" and "usbip-host.ko" modules, they should only be visible for kernel versions they exists. over time modules get renamed, merged or splitted etc

> $(PKG)_LINUX_CATEGORY:=Unstable

if you wonder about that, the menuconfig is cached: ./config/.cache.in

> libudev

just use the lib of freetz (or add it if it is not available)!
using an avm-lib causes only problems. additionally you dont know if avme adds it to all devices and/or older/newer fw versions!


this example packagehas no externel sources but it in freetz-git: https://github.com/Freetz-NG/freetz-ng/tree/master/make/host-tools/yf-akcarea-host/src

**Comment by coldfix** [comment]
Thanks for your pointers!

> maybe because of wrong libs? all freetz bins use /usr/lib/freetz/, nothing else! also freetz has own libc

Hm... the freetz-built usbip(d) binaries link against `libudev.so` and `libc.so.0`. while the avm deployed `libudev.so.1` links against `libc.so.0` and `ld-uClibc.so.1` and the latter does not export any of the same symbol names as `libc.so.0`. Also, the `libudev.h` interface does not exchange any libc library objects  (which would be prone to crashes in case of statically linked libc components), other than c-strings, but I hope they do not change ownership on these interfaces. So it doesn't seem to be the reason to me at first glance, but I might be wrong. 

> i dont know, but maybe these package help you?

as far as I can see, all of these pull in sources from externally hosted repositories, rather than make use of code that is directly in the kernel.

> yes, the old devices then no longer work. I've many packages with different versions, like grep ABANDON  make/*/*/*.mk

I see, so the goto-way is to maintain both different versions within the same `make/pkgs/<package>` folder. I will have to see whether the "new" version differs too much from the old to make this practical. In this case I would prefer to have two separate folders side-by-side.

> you added "usbip-core.ko" and "usbip-host.ko" modules, they should only be visible for kernel versions they exists. over time modules get renamed, merged or splitted etc

Yeah. I will have to check in which kernel versions they are available, and which patches are required in each version. Do you have a quick way to do this sort of task? Is there a list of all used kernel versions, so I could iterate over them?

> > `$(PKG)_LINUX_CATEGORY:=Unstable`
>
> if you wonder about that, the menuconfig is cached: ./config/.cache.in

I've rebuilt from a cleaned git repo (`git clean -xdf`) with and still the package is listed under `U` rather than `Unstable`. Weird, but not so important for now.

> just use the lib of freetz (or add it if it is not available)!
using an avm-lib causes only problems. additionally you dont know if avme adds it to all devices and/or older/newer fw versions!

On my system libudev is part of systemd which FOS does not use. I did not find a officially looking standalone upstream version online - which is why I assumed it's tightly coupled to the OS and it might be better to use the OS-shipped version. But I did not do more detailed investigations, whether/from where we could pull in and build an appropriate version ourselves. I noticed that the `make/pkgs/libusb1` and `pcsc-lite` packages disable udev. So if we could build/use a reliable version these package might also profit from it.

**Reply by fda77** [reply]
Seems libudev is part of systemd: https://github.com/systemd/systemd/tree/main/src/libudev
You could also check avm sources https://osp.avm.de
freetz supports currently 91 avm kernel, first you should only support your device...
AVM has something udev like, i have no idea if this is a problem for this!

**Reply by coldfix** [reply]
Thanks! Didn't know they release sources! That helped :) They use [eudev](https://github.com/gentoo/eudev). That should make it feasible to add a package for it (but not sure if it helps other than making our build more self-contained).

**Reply by coldfix** [reply]
But will definitely be useful for debugging purposes..

**Reply by fda77** [reply]
i found it @debian: https://packages.debian.org/de/sid/libudev-dev
